File: MiniDOMOffice/Office/RichiesteCERQ/MotiviRichieste/MotivoRichiesta.vb
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica

Partial Class Office

    <Flags> _
    Public Enum MotivoRichiestaFlags
        None = 0

    End Enum

    ''' <summary>
    ''' Rappresenta un motivo di una commissione
    ''' </summary>
    ''' <remarks></remarks>
    <Serializable> _
    Public Class MotivoRichiesta
        Inherits DBObject

        Private m_Motivo As String                      'Motivo della commissione
        Private m_Flags As MotivoRichiestaFlags
        Private m_HandlerName As String

        Public Sub New()
            Me.m_Motivo = vbNullString
            Me.m_Flags = MotivoRichiestaFlags.None
            Me.m_HandlerName = ""
        End Sub


        ''' <summary>
        ''' Restituisce o imposta il motivo della commissione (descrizione breve)
        ''' </summary>
        ''' <value></value>
        ''' <returns></returns>
        ''' <remarks></remarks>
        Public Property Motivo As String
            Get
                Return Me.m_Motivo
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_Motivo
                If (oldValue = value) Then Exit Property
                Me.m_Motivo = value
                Me.DoChanged("Motivo", value, oldValue)
            End Set
        End Property

        Public Property Flags As MotivoRichiestaFlags
            Get
                Return Me.m_Flags
            End Get
            Set(value As MotivoRichiestaFlags)
                Dim oldValue As MotivoRichiestaFlags = Me.m_Flags
                If (oldValue = value) Then Exit Property
                Me.m_Flags = value
                Me.DoChanged("Flags", value, oldValue)
            End Set
        End Property

        Public Property HandlerName As String
            Get
                Return Me.m_HandlerName
            End Get
            Set(value As String)
                Dim oldValue As String = Me.m_HandlerName
                value = Strings.Trim(value)
                If (oldValue = value) Then Exit Property
                Me.m_HandlerName = value
                Me.DoChanged("HandlerName", value, oldValue)
            End Set
        End Property

        Public Overrides Function ToString() As String
            Return Me.m_Motivo
        End Function

        Protected Overrides Function GetConnection() As CDBConnection
            Return Office.Database
        End Function

        Public Overrides Function GetModule() As Sistema.CModule
            Return Office.RichiesteCERQ.MotiviRichieste.Module
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_OfficeRichiesteM"
        End Function

        Protected Overrides Function LoadFromRecordset(reader As DBReader) As Boolean
            Me.m_Motivo = reader.Read("Motivo", Me.m_Motivo)
            Me.m_Flags = reader.Read("Flags", Me.m_Flags)
            Me.m_HandlerName = reader.Read("NomeHandler", Me.m_HandlerName)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(writer As DBWriter) As Boolean
            writer.Write("Motivo", Me.m_Motivo)
            writer.Write("Flags", Me.m_Flags)
            writer.Write("NomeHandler", Me.m_HandlerName)
            Return MyBase.SaveToRecordset(writer)
        End Function

        Protected Overrides Sub XMLSerialize(writer As XML.XMLWriter)
            writer.WriteAttribute("Motivo", Me.m_Motivo)
            writer.WriteAttribute("Flags", Me.m_Flags)
            writer.WriteAttribute("NomeHandler", Me.m_HandlerName)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(fieldName As String, fieldValue As Object)
            Select Case fieldName
                Case "Motivo" : Me.m_Motivo = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "Flags" : Me.m_Flags = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "NomeHandler" : Me.m_HandlerName = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select

        End Sub

        Protected Overrides Sub OnAfterSave(e As SystemEvent)
            MyBase.OnAfterSave(e)
            Office.RichiesteCERQ.MotiviRichieste.UpdateCached(Me)
        End Sub
    End Class



End Class

File: MiniDOMOffice/Office/RichiesteCERQ/MotiviRichieste/MotivoRichiesta.cs
using System;
using DMD;
using DMD.Databases;
using DMD.XML;
using minidom;
using static minidom.Sistema;

namespace minidom
{
    public partial class Office
    {

        /// <summary>
        /// Flag per <see cref="MotivoRichiesta"/>
        /// </summary>
        [Flags]
        public enum MotivoRichiestaFlags
        {
            /// <summary>
            /// Nessun flag
            /// </summary>
            None = 0
        }

        /// <summary>
        /// Rappresenta un motivo di una commissione
        /// </summary>
        /// <remarks></remarks>
        [Serializable]
        public class MotivoRichiesta
            : Databases.DBObject, IComparable, IComparable<MotivoRichiesta>
        {
            private string m_Motivo;                      // Motivo della commissione
            private string m_HandlerName;

            /// <summary>
            /// Costruttore
            /// </summary>
            public MotivoRichiesta()
            {
                m_Motivo = DMD.Strings.vbNullString;
                m_Flags = (int)MotivoRichiestaFlags.None;
                m_HandlerName = "";
            }


            /// <summary>
            /// Restituisce o imposta il motivo della commissione (descrizione breve)
            /// </summary>
            /// <value></value>
            /// <returns></returns>
            /// <remarks></remarks>
            public string Motivo
            {
                get
                {
                    return m_Motivo;
                }

                set
                {
                    value = Strings.Trim(value);
                    string oldValue = m_Motivo;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Motivo = value;
                    DoChanged("Motivo", value, oldValue);
                }
            }

            /// <summary>
            /// Flags
            /// </summary>
            public new MotivoRichiestaFlags Flags
            {
                get
                {
                    return (MotivoRichiestaFlags) base.Flags;
                }

                set
                {
                    var oldValue = this.Flags;
                    if (oldValue == value)
                        return;
                    m_Flags = (int)value;
                    DoChanged("Flags", value, oldValue);
                }
            }

            /// <summary>
            /// Handler
            /// </summary>
            public string HandlerName
            {
                get
                {
                    return m_HandlerName;
                }

                set
                {
                    string oldValue = m_HandlerName;
                    value = DMD.Strings.Trim(value);
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_HandlerName = value;
                    DoChanged("HandlerName", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce una stringa che rappresenta l'oggetto
            /// </summary>
            /// <returns></returns>
            public override string ToString()
            {
                return m_Motivo;
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return HashCalculator.Calculate(this.m_Motivo);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public sealed override bool Equals(Databases.DBObject obj)
            {
                return (obj is MotivoRichiesta) && this.Equals((MotivoRichiesta)obj);
            }

            /// <summary>
            /// Restituisce true se i due oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Equals(MotivoRichiesta obj)
            {
                return base.Equals(obj)
                      && DMD.Strings.EQ(this.m_Motivo, obj.m_Motivo)
                      && DMD.Strings.EQ(this.m_HandlerName, obj.m_HandlerName)
                    ;

            }

            /// <summary>
            /// Compara due oggetti
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public int CompareTo(MotivoRichiesta obj)
            {
                return DMD.Strings.Compare(this.m_Motivo, obj.m_Motivo, true);
            }

            int IComparable.CompareTo(object obj)
            {
                return this.CompareTo((MotivoRichiesta)obj);
            }

            /// <summary>
            /// Repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return minidom.Office.RichiesteCERQ.MotiviRichieste;
            }

            /// <summary>
            /// Discriminator
            /// </summary>
            /// <returns></returns>
            public override string GetTableName()
            {
                return "tbl_OfficeRichiesteM";
            }

            /// <summary>
            /// Carica dal db
            /// </summary>
            /// <param name="reader"></param>
            /// <returns></returns>
            protected override bool LoadFromRecordset(DBReader reader)
            {
                this.m_Motivo = reader.Read("Motivo", this.m_Motivo);
                this.m_HandlerName = reader.Read("NomeHandler", this.m_HandlerName);
                return base.LoadFromRecordset(reader);
            }

            /// <summary>
            /// Salva nel db
            /// </summary>
            /// <param name="writer"></param>
            /// <returns></returns>
            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("Motivo", this.m_Motivo);
                writer.Write("NomeHandler", this.m_HandlerName);
                return base.SaveToRecordset(writer);
            }

            /// <summary>
            /// Prepara lo schema
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaFields(DBTable table)
            {
                base.PrepareDBSchemaFields(table);
                var c = table.Fields.Ensure("Motivo", typeof(string), 255);
                c = table.Fields.Ensure("NomeHandler", typeof(string), 255);
                
            }

            /// <summary>
            /// Prepara i vincoli
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaConstraints(DBTable table)
            {
                base.PrepareDBSchemaConstraints(table);
                var c = table.Constraints.Ensure("idxMotivo", new string[] { "Motivo", "Stato" }, DBFieldConstraintFlags.PrimaryKey);
                c = table.Constraints.Ensure("idxHandler", new string[] { "NomeHandler" }, DBFieldConstraintFlags.None);
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("Motivo", m_Motivo);
                writer.WriteAttribute("NomeHandler", m_HandlerName);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "Motivo":
                        {
                            m_Motivo = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }
 
                    case "NomeHandler":
                        {
                            m_HandlerName = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }
             
        }
    }
}